### PR TITLE
Abstraction

### PR DESCRIPTION
**Pull request by fonglh**

### CODE REVIEW

File: chapters/2-abstraction.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# Abstraction
+
+Functional abstraction allows us to treat functions as black boxes.
+We just care about what a function does and how to use it.
```
Comment (kxmbrian) [inline chapters/2-abstraction.md]: I think the value of this paragraph is that it caters to the clueless students. The non-blur ones will get stuck on the application, not the definition here. Basically need to help them think.

1) Need to say what we do not care about. Else, students might say "okay. so we need to care about what a function does and how to use it. Isn't that everything?" "What a function does? So does the function process the values using a for loop or while loop?"

2) Maybe make the logical link explicit. "... boxes, meaning that, we just care ...". The clueless ones might not realize that you are explaining what you mean by "black boxes".

File: chapters/2-abstraction.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+**function stubs**.
+
+A [function stub](https://en.wikipedia.org/wiki/Method_stub)
+is a piece of code that stands in for the actual functionality.
+
+The stub allows us to get the main outline of the program up
+before figuring out the details.
+While incrementally working towards the final product,
+the program can be run without syntax errors, allowing us to write
+small manageable chunks at a time.
```
Comment (kxmbrian) [inline chapters/2-abstraction.md]: Show, don't tell. haha. Maybe can give a very compelling stub example, then formalize that knowledge with the definitions, properties and benefits of stubs.

File: chapters/2-abstraction.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+## Defeating the Balrog
+
+We had quite a few questions about the Balrog lecture training question.
+The main difficulties were not knowing what the question wanted,
+and unfamiliarity with Python's syntax for defining functions.
+
+### Similar Examples
+
+A closer examination of the code given in the question reveals
+two stubs: `spawn_balrog` and `balrog_attack`.
```
Comment (kxmbrian) [inline chapters/2-abstraction.md]: there's one more: take_defensive_action

"examinations.. reveals" is like saying "it is obvious that.." in math texts. Students go, "I examine and don't understand lehh.. _feels discouraged_ ".   Maybe say instead, "we have written three stubs for you already"

File: chapters/2-abstraction.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+We'll cover their main parts below.
+
+#### Function Definition
+
+Both functions start with the Python keyword `def`, for "define",
+followed by the function's name. Finally, variable names
+for the function's parameters are surrounded by parentheses.
```
Comment (kxmbrian) [inline chapters/2-abstraction.md]: got example?

not sure if there is a need to say this, unless there are students who don't already know this form the lecture + exercises. Want to keep the gitbook content as minimal as possible.

File: chapters/2-abstraction.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+Notice that both functions have an English sentence right below the `def`
+line. This is a **docstring**. It's meant to document your code,
+explaining what the function is supposed to do.
+
+[PEP 257](https://www.python.org/dev/peps/pep-0257/) explains Python's
+docstring conventions.
+
+*Why don't we just use a comment?*
+
+Docstrings are useful for documentation generators such as
```
Comment (kxmbrian) [inline chapters/2-abstraction.md]: maybe can tell them that the string will be shown when they do `help(thier_function)`. More relevant.

File: chapters/2-abstraction.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+go through your code, find the docstrings and generate various output
+formats of your documentation.
+
+By following the conventions, you can get readable documentation for
+little effort!
+
+#### Function Body
+
+In both these examples, the only code in the function is `pass`.
+The `pass` statement is needed to avoid syntax errors, but
```
Comment (kxmbrian) [inline chapters/2-abstraction.md]: I think something like, "if you leave the function body empty, then you will get a syntax error. You need to put something. You can put 'pass'. `pass` is just a placeholder that does nothing." is better.

this way, we describe how `pass` helps us avoid syntax errors.

File: chapters/2-abstraction.md (hunk 2 above)
Comment (fonglh) [inline chapters/2-abstraction.md]: The show is in the example later already. That was where they got stuck so I figured they need a more concrete explanation.

File: chapters/2-abstraction.md (hunk 4 above)
Comment (fonglh) [inline chapters/2-abstraction.md]: It's for completeness, but I'm fixing the other suggestion.

File: chapters/2-abstraction.md (hunk 5 above)
Comment (fonglh) [inline chapters/2-abstraction.md]: Added that tip.

File: chapters/2-abstraction.md (hunk 2 above)
Comment (kxmbrian) [inline chapters/2-abstraction.md]: Okay, I feel that if I were a student reading this, I wouldn't be convinced. Instead to me it comes across as a "just believe me. you eventually will. I'm the teacher". Not sure if putting this at the end helps.

Maybe I'll come up with one.

**Pull request merged**